File: sdk/src/types/aa.ts
import { z } from 'zod';

export const CreateAccountParamsSchema = z.object({
  tenantId: z.string(),
  userId: z.string(),
  ownerAddress: z.string(),
});

export type CreateAccountParams = z.infer<typeof CreateAccountParamsSchema>;

export const CreateAccountResponseSchema = z.object({
  address: z.string(),
  owner: z.string(),
  accountType: z.string(),
  isDeployed: z.boolean(),
  chainId: z.number(),
  entryPoint: z.string(),
});

export type CreateAccountResponse = z.infer<typeof CreateAccountResponseSchema>;

export const GetAccountResponseSchema = z.object({
  address: z.string(),
  isDeployed: z.boolean(),
  nonce: z.string(),
  chainId: z.number(),
  entryPoint: z.string(),
  accountType: z.string(),
});

export type GetAccountResponse = z.infer<typeof GetAccountResponseSchema>;

export const SmartAccountSchema = GetAccountResponseSchema;
export type SmartAccount = GetAccountResponse;

export const UserOperationSchema = z.object({
  sender: z.string(),
  nonce: z.string(),
  initCode: z.string().optional(),
  callData: z.string(),
  callGasLimit: z.string(),
  verificationGasLimit: z.string(),
  preVerificationGas: z.string(),
  maxFeePerGas: z.string(),
  maxPriorityFeePerGas: z.string(),
  paymasterAndData: z.string().optional(),
  signature: z.string().optional(),
});

export type UserOperation = z.infer<typeof UserOperationSchema>;

export const SendUserOperationRequestSchema = z.object({
  tenantId: z.string(),
  userOperation: UserOperationSchema,
  sponsor: z.boolean().optional(),
});

export type SendUserOperationRequest = z.infer<typeof SendUserOperationRequestSchema>;

export const SendUserOperationResponseSchema = z.object({
  userOpHash: z.string(),
  sender: z.string(),
  nonce: z.string(),
  status: z.string(),
  sponsored: z.boolean(),
});

export type SendUserOperationResponse = z.infer<typeof SendUserOperationResponseSchema>;

export const EstimateGasRequestSchema = z.object({
  tenantId: z.string(),
  userOperation: UserOperationSchema,
});

export type EstimateGasRequest = z.infer<typeof EstimateGasRequestSchema>;

export const GasEstimateSchema = z.object({
  preVerificationGas: z.string(),
  verificationGasLimit: z.string(),
  callGasLimit: z.string(),
  maxFeePerGas: z.string(),
  maxPriorityFeePerGas: z.string(),
});

export type GasEstimate = z.infer<typeof GasEstimateSchema>;

export const UserOpReceiptSchema = z.object({
  userOpHash: z.string(),
  sender: z.string(),
  nonce: z.string(),
  success: z.boolean(),
  actualGasCost: z.string(),
  actualGasUsed: z.string(),
  paymaster: z.string().optional(),
  transactionHash: z.string(),
  blockHash: z.string(),
  blockNumber: z.string(),
});

export type UserOpReceipt = z.infer<typeof UserOpReceiptSchema>;

export const SessionKeySchema = z.object({
  id: z.string().optional(),
  publicKey: z.string(),
  permissions: z.array(z.string()),
  validUntil: z.number(),
  validAfter: z.number().optional(),
});

export type SessionKey = z.infer<typeof SessionKeySchema>;

export const AddSessionKeyParamsSchema = z.object({
  accountAddress: z.string(),
  sessionKey: SessionKeySchema,
});

export type AddSessionKeyParams = z.infer<typeof AddSessionKeyParamsSchema>;

export const RemoveSessionKeyParamsSchema = z.object({
  accountAddress: z.string(),
  keyId: z.string(),
});

export type RemoveSessionKeyParams = z.infer<typeof RemoveSessionKeyParamsSchema>;

File: sdk-python/src/rampos/models/aa.py
"""Account Abstraction (ERC-4337) models."""

from __future__ import annotations

from pydantic import BaseModel


class SmartAccount(BaseModel):
    address: str
    owner: str
    is_deployed: bool
    chain_id: int
    entry_point: str
    account_type: str


class CreateAccountParams(BaseModel):
    tenant_id: str
    user_id: str
    owner_address: str


class CreateAccountResponse(BaseModel):
    address: str
    owner: str
    account_type: str
    is_deployed: bool
    chain_id: int
    entry_point: str


class UserOperation(BaseModel):
    sender: str
    nonce: str
    init_code: str | None = None
    call_data: str
    call_gas_limit: str
    verification_gas_limit: str
    pre_verification_gas: str
    max_fee_per_gas: str
    max_priority_fee_per_gas: str
    paymaster_and_data: str | None = None
    signature: str | None = None


class SendUserOperationRequest(BaseModel):
    tenant_id: str
    user_operation: UserOperation
    sponsor: bool | None = None


class SendUserOperationResponse(BaseModel):
    user_op_hash: str
    sender: str
    nonce: str
    status: str
    sponsored: bool


class EstimateGasRequest(BaseModel):
    tenant_id: str
    user_operation: UserOperation


class GasEstimate(BaseModel):
    pre_verification_gas: str
    verification_gas_limit: str
    call_gas_limit: str
    max_fee_per_gas: str
    max_priority_fee_per_gas: str


class UserOpReceipt(BaseModel):
    user_op_hash: str
    sender: str
    nonce: str
    success: bool
    actual_gas_cost: str
    actual_gas_used: str
    paymaster: str | None = None
    transaction_hash: str
    block_hash: str
    block_number: str
